### PR TITLE
BitPackedArray::scalar_at

### PR DESCRIPTION
**Pull request by lwwmanning**

### CODE REVIEW

File: fastlanez/src/transpose.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    fl_untranspose_u32, fl_untranspose_u64, fl_untranspose_u8,
+};
+
+const fn transposable<T: Sized, U: Sized>() -> bool {
+    let sizeOfT = size_of::<T>();
+    sizeOfT == size_of::<U>() && (sizeOfT == 1 || sizeOfT == 2 || sizeOfT == 4 || sizeOfT == 8)
+}
+
+pub fn transpose<T: Sized, U: Sized>(input: &[T; 1024], output: &mut [U; 1024]) {
+    assert!(
```
Comment (gatesn) [inline fastlanez/src/transpose.rs]: Can you use const generics to constrain this in the type system?

**Comment by lwwmanning** [comment]
depends on #179

File: fastlanez/src/bitpack.rs [hunk 2]
```diff
@@ -16,6 +16,11 @@ where
     Pred<{ W > 0 }>: Satisfied,
     Pred<{ W < 8 * size_of::<Self>() }>: Satisfied,
 {
+    // 1024 elements compress to W bits each -> (1024 * W / 8) bytes
+    const BYTES_PER_TRANCHE: usize = 128 * W;
```
Comment (gatesn) [inline fastlanez/src/bitpack.rs]: We don't really use the word tranche anywhere, maybe just stay consistent and inline 128 * W.

Same for below.

**Review by gatesn — DISMISSED** [review]
I think we should benchmark this. There's quite a lot of logic for 1 scalar and I think there's quite a lot of room to improve performance

**Comment by lwwmanning** [comment]
benchmark results at commit https://github.com/fulcrum-so/vortex/pull/160/commits/e88374f630091902b93a8d300197d1d9f28f90e3 on my MacBook Air M2:

```zsh
~/g/vortex ❯❯❯ cargo bench --bench bitpacking
   Compiling vortex-fastlanes v0.1.0 (/Users/will/git/vortex/vortex-fastlanes)
    Finished `bench` profile [optimized] target(s) in 1.83s
     Running benches/bitpacking.rs (target/release/deps/bitpacking-274b666c83950221)
Gnuplot not found, using plotters backend
bitpack_1M              time:   [55.673 µs 55.768 µs 55.883 µs]
Found 13 outliers among 100 measurements (13.00%)
  11 (11.00%) high mild
  2 (2.00%) high severe

unpack_1M               time:   [224.61 µs 240.60 µs 258.08 µs]
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

unpack_1M_singles       time:   [2.2627 ms 2.2864 ms 2.3125 ms]
Found 13 outliers among 100 measurements (13.00%)
  8 (8.00%) high mild
  5 (5.00%) high severe

unpack_1024             time:   [43.631 ns 43.774 ns 43.928 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

unpack_single           time:   [4.9211 ns 4.9358 ns 4.9515 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

**Review by gatesn — APPROVED** [review]

File: vortex-fastlanes/src/bitpacking/compute.rs [hunk 3]
```diff
@@ -28,6 +34,21 @@ impl FlattenFn for BitPackedArray {
...
+            return Err(vortex_err!(OutOfBounds:index, 0, self.len()));
+        }
+        if self.bit_width() == 0 {
+            let ptype = self.dtype().try_into()?;
+            match_each_integer_ptype!(&ptype, |$P| {
+                return Ok(Scalar::from(0 as $P));
+            })
+        }
+        unpack_single(self, index)
+    }
```
Comment (robert3005) [inline vortex-fastlanes/src/bitpacking/compute.rs:49]: This needs to take patches into account

**Pull request merged**